backend/layers/processing/rollback.py

```diff
@@ -0,0 +1,301 @@
+                logger.info(f"Failed to rollback DatasetVersion {dataset.version_id}, {e}")
+                continue
+        return collection_version.datasets
+
+    def rollback_private_dataset_list(self) -> None:
+        """
+        Rolls back the Datasets in the CollectionVersions associated with each DatasetVersionId passed in, to their
+        respective previous, most recently created DatasetVersion. Then, triggers deletion of the DB references and S3
+        assets for the rolled back DatasetVersions.
+        """
+        rolled_back_datasets = []
+        for dataset_version_id in self.entity_id_list:
+            dataset_version_id = DatasetVersionId(dataset_version_id)
+            try:
+                rolled_back_dataset = self.rollback_private_dataset(dataset_version_id)
+            except Exception as e:
+                logger.info(f"Failed to rollback DatasetVersion {dataset_version_id}, {e}")
+                continue
+            rolled_back_datasets.append(rolled_back_dataset)
+        self._clean_up_rolled_back_datasets(rolled_back_datasets)
+
+    def rollback_private_dataset(
+        self, dataset_version_id: DatasetVersionId, collection_version_id: CollectionVersionId = None
+    ) -> DatasetVersion:
+        """
+        For a given DatasetVersionId and unpublished CollectionVersionId, rolls back the associated Dataset in the
+        CollectionVersion to its previous, most recently created DatasetVersion and deletes the given DatasetVersion.
+
+        :param dataset_version_id: DatasetVersionId of the DatasetVersion to rollback
+        :param collection_version_id: CollectionVersionId of the CollectionVersion to rollback the DatasetVersion
+        in. If not passed in, the CollectionVersionId will be determined from the DatasetVersionId.
+        :return: DatasetVersion that was rolled back
+        """
+        cv_id = collection_version_id
+        dataset_version = self.business_logic.get_dataset_version(dataset_version_id)
+        if cv_id is None:
+            # account for collection potentially having a migration revision and a non-migration revision
+            collection_versions = self.business_logic.get_unpublished_collection_versions_from_canonical(
+                dataset_version.collection_id
+            )
+            for cv in collection_versions:
+                if cv_id is not None:
+                    break
+                for dataset in cv.datasets:
+                    if dataset.version_id == dataset_version_id:
+                        cv_id = cv.version_id
+                        break
+        if cv_id is None:
+            raise CollectionNotFoundException(
+                f"An Associated CollectionVersion not found for DatasetVersion {dataset_version_id}"
+            )
+        self.business_logic.restore_previous_dataset_version(cv_id, dataset_version.dataset_id)
+        return dataset_version
+
+    def rollback_public_collections(self) -> None:
```

Comment:
It looks like this could use `rollback_public_collection_list` and remove some duplicated logic.